### PR TITLE
RDM-8675: Implement BEFTA Functional Tests for 'Get Case-Assigned Users and Roles'

### PR DESCRIPTION
**Pull request by kiran-yenigala-hmcts**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/RDM-8675

### Change description ###

Functional Test cases added to address RDM-8344

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X ] No
```

### CODE REVIEW

File: src/aat/resources/features/F-103/F-103.feature [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+Feature: F-103: Get Case-Assigned Users and Roles
+
+  Background: Load test data for the scenario
+    Given an appropriate test context as detailed in the test data source
+
+  @S-597
+  Scenario: when a specific User ID is supplied for a specific case ID, then the case roles relating only to the User ID for that specific Case ID must be returned
+    Given an appropriate test context as detailed in the test data source,
+    And a user [Richard - who can create a case],
+    And a successful call [to create a token for case creation] as in [F103_GetToken]
```
Comment (MSancaktutar) [inline src/aat/resources/features/F-103/F-103.feature:11]: pls remove token creation lines

File: src/aat/resources/features/F-103/F-103_Test_Case_Access_Base_Dil_C2.td.json [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+      "username": "befta.caseworker.caa.1@gmail.com",
+      "password": "[[$CCD_CASEWORKER_AUTOTEST_PASSWORD]]"
+    },
+    "userRichard": {
+      "username": "befta.caseworker.2.solicitor.2@gmail.com",
+      "password": "[[$CCD_CASEWORKER_AUTOTEST_PASSWORD]]"
+    },
+    "userDil": {
+      "username": "befta.caseworker.2.solicitor.3@gmail.com",
+      "password": "[[$CCD_CASEWORKER_AUTOTEST_PASSWORD]]"
```
Comment (MSancaktutar) [inline src/aat/resources/features/F-103/F-103_Test_Case_Access_Base_Dil_C2.td.json:16]: each account has it's respective password env var. please always use respective env vars.

Comment (MSancaktutar) [reply src/aat/resources/features/F-103/F-103_Test_Case_Access_Base_Dil_C2.td.json:16]: in general...

File: src/aat/resources/features/F-103/F-103_Test_Case_Access_Base_Dil_C3.td.json [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+      "password": "[[$CCD_CASEWORKER_AUTOTEST_PASSWORD]]"
+    },
+    "userDil": {
+      "username": "befta.caseworker.2.solicitor.3@gmail.com",
+      "password": "[[$CCD_CASEWORKER_AUTOTEST_PASSWORD]]"
+    },
+    "userSteve": {
+      "username": "befta.caseworker.2.solicitor.4@gmail.com",
+      "password": "[[$CCD_CASEWORKER_AUTOTEST_PASSWORD]]"
+    }
```
Comment (MSancaktutar) [inline src/aat/resources/features/F-103/F-103_Test_Case_Access_Base_Dil_C3.td.json:21]: This lines seem to be unnecessarily repeated. Am I missing a detail?

Comment (MSancaktutar) [reply src/aat/resources/features/F-103/F-103_Test_Case_Access_Base_Dil_C3.td.json:21]: JSON-based inheritance mechanism would already bring the same lines from a parent, base object.

File: src/aat/resources/features/F-103/F-103_Test_Case_Access_Base_Steve.td.json [hunk 4]
```diff
@@ -0,0 +1,30 @@
+{
+	"_guid_": "F-103_Test_Case_Access_Base_Steve",
+	"method": "PUT",
+	"uri": "/cases/{cid}/users/{uid}",
```
Comment (MSancaktutar) [inline src/aat/resources/features/F-103/F-103_Test_Case_Access_Base_Steve.td.json:4]: there should be an _extends_ line here to reuse content of the common, base test data object.

File: src/aat/resources/features/F-103/F-103_Test_Case_Access_Base_Steve_C3.td.json [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+      "password": "[[$CCD_CASEWORKER_AUTOTEST_PASSWORD]]"
+    },
+    "userDil": {
+      "username": "befta.caseworker.2.solicitor.3@gmail.com",
+      "password": "[[$CCD_CASEWORKER_AUTOTEST_PASSWORD]]"
+    },
+    "userSteve": {
+      "username": "befta.caseworker.2.solicitor.4@gmail.com",
+      "password": "[[$CCD_CASEWORKER_AUTOTEST_PASSWORD]]"
+    }
```
Comment (MSancaktutar) [inline src/aat/resources/features/F-103/F-103_Test_Case_Access_Base_Steve_C3.td.json]: in general, most of these repeated lines could be avoided by inheriting them from appropriate base objects. pls could you revise all test data files from this perspective, to maximise re-use and minimise repeated test data lines?

File: src/aat/resources/features/F-103/S-603.td.json [hunk 6]
```diff
@@ -0,0 +1,36 @@
+{
+  "title": "must return 400 error response for mal-formatted user id list",
+  "_guid_": "S-603",
```
Comment (MSancaktutar) [inline src/aat/resources/features/F-103/S-603.td.json:3]: guid repeated in multiple files. that will mess things.

File: src/aat/resources/features/F-103/S-600.td.json [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    "Dil - with an active profile",
+    "Steve - with an active profile",
+    "Jamal -  who is a privileged user with permissions to access the case assignments of other users",
+    "by Jamal to assign Dil a few case roles to access C1",
+    "by Jamal to assign Dil a few case roles to access C2",
+    "by Jamal to assign Dil a few case roles to access C3",
+    "by Jamal to assign Steve a few case roles to access C1",
+    "by Jamal to assign Steve a few case roles to access C2",
+    "by Jamal to assign Steve a few case roles to access C3",
+    "is made by Jamal with Case IDs of C1, C2 & C3 & no User ID"
```
Comment (MSancaktutar) [inline src/aat/resources/features/F-103/S-600.td.json:19]: are you sure you will have all these specs here? they look like specs to go to the respective api call test data.

Comment (MSancaktutar) [reply src/aat/resources/features/F-103/S-600.td.json:19]: Please check other test data files to confirm you have all and only the necessary specs - no more, no less. Less than enough fails test, more than enough makes the maintenance quite confusing later.

File: src/aat/resources/features/F-103/S-603.td.json [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    },
+    "body": {
+      "exception": "uk.gov.hmcts.ccd.endpoint.exceptions.BadRequestException",
+      "timestamp": "[[ANYTHING_PRESENT]]",
+      "message": "User ID is not valid",
+      "path": "[[ANYTHING_PRESENT]]",
+      "details": null,
+      "callbackErrors": null,
+      "callbackWarnings": null
+    }
```
Comment (MSancaktutar) [inline src/aat/resources/features/F-103/S-603.td.json:34]: when you have a section of the test data inheriting from Common_XYZ_Response, you won't need to repeat many lines like these, as they will already be inherited from the base object

File: src/aat/resources/features/F-103/S-603.td.json (hunk 6 above)
Comment (kiran-yenigala-hmcts) [reply src/aat/resources/features/F-103/S-603.td.json:3]: fixed this!!!! Copy paste error!!

File: src/aat/resources/features/F-103/S-603.td.json (hunk 8 above)
Comment (kiran-yenigala-hmcts) [reply src/aat/resources/features/F-103/S-603.td.json:34]: Changed this, Pleaser review!!!

File: src/aat/resources/features/F-103/F-103_Test_Case_Access_Base_Steve.td.json (hunk 4 above)
Comment (kiran-yenigala-hmcts) [reply src/aat/resources/features/F-103/F-103_Test_Case_Access_Base_Steve.td.json:4]: Done

File: src/aat/resources/features/F-103/F-103_Test_Case_Access_Base_Dil_C3.td.json (hunk 3 above)
Comment (kiran-yenigala-hmcts) [reply src/aat/resources/features/F-103/F-103_Test_Case_Access_Base_Dil_C3.td.json]: Done, used extends!!!

File: src/aat/resources/features/F-103/F-103_Test_Case_Access_Base_Dil_C2.td.json (hunk 2 above)
Comment (kiran-yenigala-hmcts) [reply src/aat/resources/features/F-103/F-103_Test_Case_Access_Base_Dil_C2.td.json:16]: Used user specific pwd, and added new entry for befta.caseworker.caa@gmail.com

File: src/aat/resources/features/F-103/F-103.feature (hunk 1 above)
Comment (kiran-yenigala-hmcts) [reply src/aat/resources/features/F-103/F-103.feature:11]: removed!!

File: src/aat/resources/features/F-103/F-103_Test_Case_Access_Base_Steve_C3.td.json (hunk 5 above)
Comment (kiran-yenigala-hmcts) [reply src/aat/resources/features/F-103/F-103_Test_Case_Access_Base_Steve_C3.td.json]: done!!

File: src/aat/resources/features/F-103/S-600.td.json (hunk 7 above)
Comment (kiran-yenigala-hmcts) [reply src/aat/resources/features/F-103/S-600.td.json]: removed these from main spec

File: src/aat/resources/features/F-103/S-603.td.json (hunk 6 above)
Comment (kiran-yenigala-hmcts) [reply src/aat/resources/features/F-103/S-603.td.json:3]: Fixed it!!

File: src/aat/resources/features/F-103/F-103.feature [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+    And a successful call [by Richard to create a case - C3] as in [F103_Case_Data_Create_C3],
+    And a user [Dil - with an active profile],
+    And a user [Steve - with an active profile],
+    And a user [Jamal -  who is a privileged user with permissions to access the case assignments of other users],
+    And a successful call [by Jamal to assign Dil a few case roles to access C1] as in [F-103_Test_Case_Access_Base_Dil],
+    And a successful call [by Jamal to assign Dil a few case roles to access C2] as in [F-103_Test_Case_Access_Base_Dil_C2],
+    And a successful call [by Jamal to assign Dil a few case roles to access C3] as in [F-103_Test_Case_Access_Base_Dil_C3],
+    And a successful call [by Jamal to assign Steve a few case roles to access C1] as in [F-103_Test_Case_Access_Base_Steve],
+    And a successful call [by Jamal to assign Steve a few case roles to access C2] as in [F-103_Test_Case_Access_Base_Steve_C2],
+    And a successful call [by Jamal to assign Steve a few case roles to access C3] as in [F-103_Test_Case_Access_Base_Steve_C3],
```
Comment (MSancaktutar) [inline src/aat/resources/features/F-103/F-103.feature:69]: Could we discuss these specs and test data id's?

**Pull request merged**